[Emotion] Convert EuiToken

miukimiu opened the pull request:
### Summary

This PR converts **EuiToken** to Emotion.

- I also improved a few comments in the code and in the props table. Some of them in my opinion were difficult to understand.
- When in size  `xs` and `square` shape the icon was touching the top and bottom borders so I added a small vertical padding (1px)
- Added a `border-radius `to the `rectangle` shape just because I think it looks better and it gets similar to the `square` shape

<img width="1172" alt="before-after-demo@2x" src="https://user-images.githubusercontent.com/2750668/180840865-9b96af8f-7df7-45ab-ace9-98cce4ef596e.png">


This is only for testing purposes:

<img width="1162" alt="before-after@2x" src="https://user-images.githubusercontent.com/2750668/180849901-5869c8f9-dd65-41a5-af77-a454b9c907c6.png">


### Fixed EuiIconTokenStruct

Why isn't the token getting centered? Well... It seems that it was incorrectly exported from Figma/Sketch. So I fixed the design and recompiled the icon.  

<img width="1411" alt="tokenStruct@2x" src="https://user-images.githubusercontent.com/2750668/180829138-13e2d718-14f1-4e17-9700-2ee8c72880c0.png">

### Changed tokens that were defaulting to `rectangle` shape

I didn't see any rectangle shape usage. So I decided to make these tokens more aligned with the other tokens.

<img width="427" alt="defaults@2x" src="https://user-images.githubusercontent.com/2750668/180842447-86a1e8f7-ea72-4ec9-85c4-b1f9cbd2ae8f.png">



### Checklist

- [x] Checked in both **light and dark** modes
- ~[ ] Checked in **mobile**~
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- ~[ ] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#adding-playground-toggles)**~
- [x] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**
- ~[ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples~
- [x] Added or updated **[jest](https://github.com/elastic/eui/blob/main/wiki/testing.md) and [cypress](https://github.com/elastic/eui/blob/main/wiki/cypress-testing.md) tests**
- [x] Checked for **breaking changes** and labeled appropriately
- ~[ ] Checked for **accessibility** including keyboard-only and screenreader modes~
- [x] Updated the **[Figma](https://www.figma.com/community/file/964536385682658129)** library counterpart
- [x] A **[changelog](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

miukimiu commented on src/components/token/token.tsx:
```diff
@@ -7,70 +7,38 @@
 import classNames from 'classnames';
-import { CommonProps, keysOf } from '../common';
-import { isColorDark, hexToRgb } from '../../services';
+import defaults from 'lodash/defaults';
```
I wanted to get rid of this `defaults` but my attempts to replace it with pure JS failed. 😢

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

miukimiu commented:
Thanks, @constancecchen for pushing https://github.com/elastic/eui/pull/6067/commits/9ec7d9cd66cb5b0d75e8719a4ed305153e59132a. I tested and works perfectly! 👍🏽

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

chandlerprall commented on src/components/token/token.styles.ts:
```diff
@@ -0,0 +1,150 @@
+    color: ${iconColor};
+  
+
+    &[class*='-light'] {
```
Are these attribute selectors necessary? If I'm reading this correctly, since color mode is passed in the style generation should know which form is needed and can return just those styles.

miukimiu replied:
When I started doing this conversion I also got confused with these `light` and `dark` props and class names. They correspond to the **fills**. 

<img width="691" alt="Screenshot 2022-07-27 at 22 06 02" src="https://user-images.githubusercontent.com/2750668/181372430-1b99bb63-f0b6-43a4-9d60-ad5cf69f0b5e.png">

It is not related to the color mode. When a token is `light` it looks like this:

<img width="75" alt="Screenshot 2022-07-27 at 22 15 33" src="https://user-images.githubusercontent.com/2750668/181373814-d38fa460-646f-4f04-bdb8-1e071cfee515.png">

When is `dark`:

<img width="71" alt="Screenshot 2022-07-27 at 22 14 49" src="https://user-images.githubusercontent.com/2750668/181373871-aff1faaf-617c-46e0-af41-a13438e46088.png">

When is `none`:

<img width="39" alt="Screenshot 2022-07-27 at 22 17 48" src="https://user-images.githubusercontent.com/2750668/181374189-796d87e7-9804-48ae-99e6-2909a30e95f6.png">

chandlerprall replied:
How do you feel about passing the `fill` prop through, keeping the css tied to the prop value instead of matching against the generated selector. Looks like it would require creating a `token.types.ts` file to avoid a circular dependency.

```
    ${
      fill === 'light'
        ? `
          color: ${lightColor};
          background-color: ${backgroundLightColor};
          box-shadow: inset 0 0 0 1px ${boxShadowColor};
        `
        : `
          color: ${darkColor};
          background-color: ${backgroundDarkColor};
        `
    }
```

chandlerprall commented on src/components/token/token.tsx:
```diff
@@ -79,8 +46,8 @@ export interface TokenProps {
    * For best results use one of the vis color names (or 'gray').
-   * Or supply your own color (can be used with dark or no fill only).
-   * Default: `gray`
+   * Or supply your own HEX color (can be used with fill 'dark' or fill 'none' only).
```
Passing `color` appears to have the same effect when `fill="light"`; I see finalFill is forced to dark in this case, but the comment here doesn't look accurate

chandlerprall commented on src/components/token/token.tsx:
```diff
@@ -138,58 +105,60 @@ export const EuiToken: FunctionComponent<EuiTokenProps> = ({
+  const euiTheme = useEuiTheme();
+  const styles = euiTokenStyles(euiTheme);
+
+  const isTokenColor = COLORS.includes(finalColor as TokenColor);
```
We can make this a little more typescript-esque to remove the need for `finalColor as TokenColor` later in the file. Converting `isTokenColor` to a function that lives outside the component,

```typescript
const isTokenColor = (color: string): color is TokenColor =>
  COLORS.includes(color as TokenColor);
```

and then using it in the if condition,

```typescript
  if (isTokenColor(finalColor)) {
    cssStyles = [...cssStyles, styles[finalColor]];
```

let's TS keep up with the type of `finalColor` within the if block. Removing the `as TokenColor` there helps future proof the logic against potential updates to TokenColor that aren't accounted for in the isTokenColor check (unlikely to happen, but possible).

chandlerprall reviewed (changes requested):
Removing the `rectangle` shape is a breaking change. Have a couple very small requests, otherwise this LGTM

miukimiu commented:
Thanks, @chandlerprall. I addressed your review with https://github.com/elastic/eui/pull/6067/commits/720c0d4ad5c4ac2387c96de3fefc8a9fcd604050. Let me know if is there anything else to address.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

miukimiu commented on upcoming_changelogs/6067.md:
```diff
@@ -0,0 +1,7 @@
+
+**Breaking changes**
+
+- Updated `tokenFile`, `tokenSymbol` and `tokenRepo` default shapes to `square` instead of `rectangle`
```
The `rectangle` is still an option for all `EuiToken`s:

<img width="509" alt="Screenshot 2022-07-28 at 17 43 52" src="https://user-images.githubusercontent.com/2750668/181592763-96903b33-4f7b-4dbe-aec5-ecc5a89193db.png">


But I'm just updating the way the tokens `tokenFile`, `tokenSymbol`, and `tokenRepo` render by default. As we can see on the following screenshot they would render as rectangles and they looked very different from the other tokens. But it is still possible to change their shapes to `rectangles`. 

<img width="247" alt="Screenshot 2022-07-28 at 17 39 34" src="https://user-images.githubusercontent.com/2750668/181592284-466dc960-a651-468b-ab9d-5198bd9a3a25.png">


Not sure I'm explaining this clearly. 😄

chandlerprall replied:
The current changes include

```diff
- type TokenShape = 'circle' | 'square' | 'rectangle';
+ export const SHAPES = ['circle', 'square'] as const;
+ export type TokenShape = typeof SHAPES[number];
```

removing the rectangle option for consumers,

<img width="671" alt="Screen Shot 2022-07-28 at 11 01 31 AM" src="https://user-images.githubusercontent.com/313125/181596167-833ff937-cec4-4662-96a0-40abba59fbed.png">

<img width="788" alt="Screen Shot 2022-07-28 at 11 01 41 AM" src="https://user-images.githubusercontent.com/313125/181596176-359fe695-a025-4dfc-b4d5-2686ec1e76c1.png">

But sounds like this is unintentional, adding `rectangle` to the SHAPES array should resolve and will make this non-breaking again.

miukimiu replied:
Yes, it was intentional. 🙈

Thanks for catching this! I pushed https://github.com/elastic/eui/pull/6067/commits/f3859636c1ea95051e96c550a00b2032f6e289ea.

At first I decided to remove the `rectangle`. Then I didn't want to introduce a breaking change. But it seems my localhost prop table was still showing the `reactangle`. And the only example I have with a rectangle is `src-docs/src/views/icon/custom_tokens.js` so no TS complains.  

I'm going to update the docs examples to TS to prevent these mistakes. 😄

chandlerprall replied:
> Then I didn't want to introduce a breaking change. But it seems my localhost prop table was still showing the `reactangle`

FWIW this is caused by webpack caching the typescript->proptype conversion and it didn't propagate a change through that cache. I'm not sure when it happens, but making a change to the file(s) containing the types and components should clear it out. It's a limitation where the custom babel loader is unable to tell webpack about additional dependencies.

Long story short: not your fault, it's a state our dev server can sometimes find itself in.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6067/

miukimiu commented:
Thanks @chandlerprall, 

-  I'm now passing the fill to `getTokenColor()` https://github.com/elastic/eui/pull/6067/commits/0b3f9f63d332462b40c43a367525559a152734b4
- Added the `reactangle` shape back to the `TokenShape` type https://github.com/elastic/eui/pull/6067/commits/f3859636c1ea95051e96c550a00b2032f6e289ea 
- Renamed and converted the examples to TS https://github.com/elastic/eui/pull/6067/commits/6eeff256aae5238501c5f811df3db171e96560b7

chandlerprall reviewed (approved):
Changes LGTM; I don't see any visual differences, including light+dark mode with various fills & colors
